### ISSUE TITLE
Runtime crashes for check-expr

### ISSUE THREAD

**Issue by SandaruJayawardana** [ballerina-platform/ballerina-lang#35150]
**Description:**
$title.

**Steps to reproduce:**
```ballerina
function fn2() returns error? {
    boolean b = true;
    error err = error("a");
    error a1 = check b? err:err;
}
```

Output:
```
[2022-02-21 17:23:37,992] SEVERE {b7a.log.crash} - Bad type on operand stack
Exception Details:
  Location:
    test2.fn2(Lio/ballerina/runtime/internal/scheduling/Strand;)Ljava/lang/Object; @249: ifgt
  Reason:
    Type 'java/lang/Object' (current frame, stack[0]) is not assignable to integer
  Current Frame:
    bci: @249
    flags: { }
    locals: { 'io/ballerina/runtime/internal/scheduling/Strand', integer, 'io/ballerina/runtime/internal/values/ErrorValue', 'io/ballerina/runtime/api/values/BString', 'io/ballerina/runtime/internal/values/ErrorValue', null, 'io/ballerina/runtime/internal/values/TypedescValueImpl', 'java/lang/Object', 'java/lang/Object', 'io/ballerina/runtime/internal/values/ErrorValue', 'io/ballerina/runtime/internal/values/ErrorValue', 'java/lang/Object', 'java/lang/Object', integer }
    stack: { 'java/lang/Object' }
  Bytecode:
    0000000: 2ab4 0011 9900 07b8 0017 bf03 3c01 4d01
    0000010: 4e01 3a04 013a 0501 3a06 013a 0701 3a08
    0000020: 013a 0901 3a0a 013a 0b01 3a0c 0336 0d2a
    0000030: b400 1b9d 0101 150d ab00 0000 0000 0150
    0000040: 0000 0008 0000 0000 0000 004c 0000 0001
    0000050: 0000 008c 0000 0002 0000 00ba 0000 0003
    0000060: 0000 00ca 0000 0004 0000 00d5 0000 0005
    0000070: 0000 00e4 0000 0006 0000 00ec 0000 0007
    0000080: 0000 00f4 1100 0036 0d12 1c3c b200 224e
    0000090: 013a 0519 05c0 0024 3a04 bb00 2659 bb00
    00000a0: 2859 b200 2e03 b700 3110 00bd 0033 b700
    00000b0: 363a 0619 062a 1400 3788 bd00 3ab9 0040
    00000c0: 0300 3a07 1100 0136 0d2a 1907 04b8 0046
    00000d0: 3a08 2ab6 004a 9a00 b2a7 0003 bb00 2459
    00000e0: b200 502d 1904 1908 b700 534d 1bb8 0059
    00000f0: 3a0b 1100 0236 0d19 0b9d 0006 a700 0e2c
    0000100: 3a0a 1100 0336 0da7 000e 2c3a 0a11 0004
    0000110: 360d a700 0319 0a3a 0901 3a0c 1100 0536
    0000120: 0da7 000b 1100 0636 0da7 0003 1100 0736
    0000130: 0d19 0cb0 2ab4 005d 2a59 b400 1b04 645a
    0000140: b500 1b32 c000 5f59 b400 633a 0c59 b400
    0000150: 663c 59b4 006a 4d59 b400 6d4e 59b4 0070
    0000160: 3a04 59b4 0074 3a07 59b4 0077 3a08 59b4
    0000170: 007a 3a09 59b4 007d 3a0a 59b4 0080 3a0b
    0000180: b400 8336 0da7 feb1 bb00 5f59 b700 8459
    0000190: 190c b500 6359 1bb5 0066 592c b500 6a59
    00001a0: 2db5 006d 5919 04b5 0070 5919 07b5 0074
    00001b0: 5919 08b5 0077 5919 09b5 007a 5919 0ab5
    00001c0: 007d 5919 0bb5 0080 5915 0db5 0083 3a0e
    00001d0: 2ab4 005d 2a59 b400 1b5a 0460 b500 1b19
    00001e0: 0e53 190c b0                           
  Stackmap Table:
    same_frame(@11)
    full_frame(@54,{Object[#13],Integer,Object[#36],Object[#139],Object[#36],Null,Null,Object[#51],Object[#51],Object[#36],Object[#36],Object[#4],Object[#4],Integer},{})
    same_frame_extended(@132)
    full_frame(@196,{Object[#13],Integer,Object[#36],Object[#139],Object[#36],Null,Object[#38],Object[#4],Object[#51],Object[#36],Object[#36],Object[#4],Object[#4],Integer},{})
    full_frame(@220,{Object[#13],Integer,Object[#36],Object[#139],Object[#36],Null,Object[#38],Object[#4],Object[#4],Object[#36],Object[#36],Object[#4],Object[#4],Integer},{})
    same_frame(@242)
    same_frame(@255)
    same_frame(@258)
    same_frame(@266)
    same_frame(@269)
    same_frame(@277)
    same_frame(@284)
    full_frame(@292,{Object[#13],Integer,Object[#36],Object[#139],Object[#36],Null,Null,Object[#51],Object[#51],Object[#36],Object[#36],Object[#4],Object[#4],Integer},{})
    full_frame(@300,{Object[#13],Integer,Object[#36],Object[#139],Object[#36],Null,Object[#38],Object[#4],Object[#4],Object[#36],Object[#36],Object[#4],Object[#4],Integer},{})
    full_frame(@308,{Object[#13],Integer,Null,Null,Null,Null,Null,Null,Null,Null,Null,Null,Null,Integer},{})
    full_frame(@392,{Object[#13],Integer,Object[#36],Object[#139],Object[#36],Null,Object[#38],Object[#4],Object[#4],Object[#36],Object[#36],Object[#4],Object[#4],Integer},{})
 
java.lang.VerifyError: Bad type on operand stack
Exception Details:
  Location:
    test2.fn2(Lio/ballerina/runtime/internal/scheduling/Strand;)Ljava/lang/Object; @249: ifgt
  Reason:
    Type 'java/lang/Object' (current frame, stack[0]) is not assignable to integer
  Current Frame:
    bci: @249
    flags: { }
    locals: { 'io/ballerina/runtime/internal/scheduling/Strand', integer, 'io/ballerina/runtime/internal/values/ErrorValue', 'io/ballerina/runtime/api/values/BString', 'io/ballerina/runtime/internal/values/ErrorValue', null, 'io/ballerina/runtime/internal/values/TypedescValueImpl', 'java/lang/Object', 'java/lang/Object', 'io/ballerina/runtime/internal/values/ErrorValue', 'io/ballerina/runtime/internal/values/ErrorValue', 'java/lang/Object', 'java/lang/Object', integer }
    stack: { 'java/lang/Object' }
  Bytecode:
    0000000: 2ab4 0011 9900 07b8 0017 bf03 3c01 4d01
    0000010: 4e01 3a04 013a 0501 3a06 013a 0701 3a08
    0000020: 013a 0901 3a0a 013a 0b01 3a0c 0336 0d2a
    0000030: b400 1b9d 0101 150d ab00 0000 0000 0150
    0000040: 0000 0008 0000 0000 0000 004c 0000 0001
    0000050: 0000 008c 0000 0002 0000 00ba 0000 0003
    0000060: 0000 00ca 0000 0004 0000 00d5 0000 0005
    0000070: 0000 00e4 0000 0006 0000 00ec 0000 0007
    0000080: 0000 00f4 1100 0036 0d12 1c3c b200 224e
    0000090: 013a 0519 05c0 0024 3a04 bb00 2659 bb00
    00000a0: 2859 b200 2e03 b700 3110 00bd 0033 b700
    00000b0: 363a 0619 062a 1400 3788 bd00 3ab9 0040
    00000c0: 0300 3a07 1100 0136 0d2a 1907 04b8 0046
    00000d0: 3a08 2ab6 004a 9a00 b2a7 0003 bb00 2459
    00000e0: b200 502d 1904 1908 b700 534d 1bb8 0059
    00000f0: 3a0b 1100 0236 0d19 0b9d 0006 a700 0e2c
    0000100: 3a0a 1100 0336 0da7 000e 2c3a 0a11 0004
    0000110: 360d a700 0319 0a3a 0901 3a0c 1100 0536
    0000120: 0da7 000b 1100 0636 0da7 0003 1100 0736
    0000130: 0d19 0cb0 2ab4 005d 2a59 b400 1b04 645a
    0000140: b500 1b32 c000 5f59 b400 633a 0c59 b400
    0000150: 663c 59b4 006a 4d59 b400 6d4e 59b4 0070
    0000160: 3a04 59b4 0074 3a07 59b4 0077 3a08 59b4
    0000170: 007a 3a09 59b4 007d 3a0a 59b4 0080 3a0b
    0000180: b400 8336 0da7 feb1 bb00 5f59 b700 8459
    0000190: 190c b500 6359 1bb5 0066 592c b500 6a59
    00001a0: 2db5 006d 5919 04b5 0070 5919 07b5 0074
    00001b0: 5919 08b5 0077 5919 09b5 007a 5919 0ab5
    00001c0: 007d 5919 0bb5 0080 5915 0db5 0083 3a0e
    00001d0: 2ab4 005d 2a59 b400 1b5a 0460 b500 1b19
    00001e0: 0e53 190c b0                           
  Stackmap Table:
    same_frame(@11)
    full_frame(@54,{Object[#13],Integer,Object[#36],Object[#139],Object[#36],Null,Null,Object[#51],Object[#51],Object[#36],Object[#36],Object[#4],Object[#4],Integer},{})
    same_frame_extended(@132)
    full_frame(@196,{Object[#13],Integer,Object[#36],Object[#139],Object[#36],Null,Object[#38],Object[#4],Object[#51],Object[#36],Object[#36],Object[#4],Object[#4],Integer},{})
    full_frame(@220,{Object[#13],Integer,Object[#36],Object[#139],Object[#36],Null,Object[#38],Object[#4],Object[#4],Object[#36],Object[#36],Object[#4],Object[#4],Integer},{})
    same_frame(@242)
    same_frame(@255)
    same_frame(@258)
    same_frame(@266)
    same_frame(@269)
    same_frame(@277)
    same_frame(@284)
    full_frame(@292,{Object[#13],Integer,Object[#36],Object[#139],Object[#36],Null,Null,Object[#51],Object[#51],Object[#36],Object[#36],Object[#4],Object[#4],Integer},{})
    full_frame(@300,{Object[#13],Integer,Object[#36],Object[#139],Object[#36],Null,Object[#38],Object[#4],Object[#4],Object[#36],Object[#36],Object[#4],Object[#4],Integer},{})
    full_frame(@308,{Object[#13],Integer,Null,Null,Null,Null,Null,Null,Null,Null,Null,Null,Null,Integer},{})
    full_frame(@392,{Object[#13],Integer,Object[#36],Object[#139],Object[#36],Null,Object[#38],Object[#4],Object[#4],Object[#36],Object[#36],Object[#4],Object[#4],Integer},{})

        at $_init.$lambda$main$(.)
        at io.ballerina.runtime.internal.scheduling.SchedulerItem.execute(Scheduler.java:611)
        at io.ballerina.runtime.internal.scheduling.Scheduler.run(Scheduler.java:317)
        at io.ballerina.runtime.internal.scheduling.Scheduler.runSafely(Scheduler.java:284)
        at java.base/java.lang.Thread.run(Thread.java:834)
```

**Affected Versions:**
SL 2201.0.0

**Comment by HindujaB** [comment]
In the above sample, I am assuming that the intention of `check` is to early return for the errors returning inside the `err:err` part. In that case, we might need to have a `check` for each case. 

Eg : if we have function calls instead of `err` parts.
```ballerina
int intResult = b ? check func(6) : check func(3);
function func(int a) returns int {
    if a > 5 {
        panic error("error");
    }
    return a + 3;
}
```
And, for the following statement, 
```ballerina
    error a1 = check b? err:err;
```
the `check` expression is considered only for the `b` part of the Elvis operator because the precedence of the `check` expression is higher. And we can't have brackets with the `check` to change the precedence as well. (Eg:`error a1 = check (b ? err : err);` won't work) 
IMO, this should be a compilation error because `check b` is not a valid expression since `b` does not return any error.

**Comment by MaryamZi** [comment]
> IMO, this should be a compilation error because check b is not a valid expression since b does not return any error.

This is valid according to the spec. Giving a compilation error was in fact the previous behaviour in jBallerina, but this was relaxed - see https://github.com/ballerina-platform/ballerina-lang/issues/33717, https://github.com/ballerina-platform/nballerina/pull/680#discussion_r745484152.

**Comment by HindujaB** [comment]
I agree @MaryamZi. 

Then still, we might need to handle this at the desugaring level. 
```ballerina
error a1 = check b? err:err;
```
The `check b` expression here will be desugared into `<boolean|error> b` but it is expected as a `boolean` inside the ternery operator. Therefore, it needs to be casted into `boolean` in the above case.  

When generating the code from BIR, the desugared `if` condition contains a ` <boolean|error>` which is wrong and will fail at compilation for the normal cases. 
```ballerina
boolean|error c = check func();
    if (c) { // compilation error
        a1 = err;
    }
```

**Comment by lochana-chathura** [comment]
Fixed in #35928
